### PR TITLE
fix unintended early return in ondemand unit test loops

### PR DESCRIPTION
**Pull request by pauldreik**
Seems like the unit test in ondemand looped, but returned on the first invocation instead of returning only on errors.

- there were a lot of warnings appearing in qtcreator with signed/unsigned/type mismatch so I tried to fix those.
- qualify with std::
- abort on error

### CODE REVIEW

**Comment by pauldreik** [comment]
Needs review by @lemire or @jkeiser

**Comment by lemire** [comment]
Just for context... who wrote the original code you are modifying?

**Comment by pauldreik** [comment]
Seems to be originally written by John, then modified by you.

**Comment by lemire** [comment]
Let us give @jkeiser a chance. Should he delay, I will study this code. I've honestly not read it carefully.

File: tests/ondemand/ondemand_basictests.cpp
Comment (jkeiser) [inline tests/ondemand/ondemand_basictests.cpp:194]: How necessary is it to pull the lambda out of the call? Most of the other tests leave it as part of the test_ondemand call, so this will look a tiny bit out of place.

File: tests/ondemand/ondemand_basictests.cpp
Comment (jkeiser) [inline tests/ondemand/ondemand_basictests.cpp:183]: I believe there is a copy of this function in tests/basictests.cpp, yes? I didn't think I'd modified it, so the same change might need to be made there.

**Comment by jkeiser** [comment]
@pauldreik I can't find where you've changed behavior, but your description seems to indicate there was an issue with a loop returning early? Can you point me at where that problem was and how it got fixed?

I am fine with the changes generally, except I'd prefer we not do the lambda assignment if we can help it, for consistency's sake. And if you don't mind checking tests/basictests.cpp for the same tests, it'd be worth keeping them in sync--I copied because we are using different types, but DOM and On Demand code is nearly identical. basictests.cpp doesn't block this checkin though :)

**Comment by pauldreik** [comment]
On old line 315, function powers_of_ten(), the loop was

for(i.....) {
   return test_ondemand( .......);
} // i

which is now

for(i.....) {
   auto test=[&].......;
   if(!test)
     return false;
   }
   return true;
} // i

The same pattern is found at powers_of_two(), small_integers()

Sorry I mixed this with the general const and std:: cleanup, I tried to get your attention on commit in a previous pull request where I believe it was easier to see.

File: tests/ondemand/ondemand_basictests.cpp
Comment (pauldreik) [reply tests/ondemand/ondemand_basictests.cpp:194]: I think it made it easier to read (nesting level) - but sure, could be rewritten. Do you think the new version makes it easier to spot the (erroneous) early return?

File: tests/ondemand/ondemand_basictests.cpp
Comment (pauldreik) [reply tests/ondemand/ondemand_basictests.cpp:183]: Ok, maybe it should be pulled out to a separate file in that case?

Comment (pauldreik) [reply tests/ondemand/ondemand_basictests.cpp:183]: Found three copies, extracted them to a separate file and added a unit test for it. But I think it looks suspicious that the same constant 0x800... is used for both the float and double version?

File: tests/ondemand/ondemand_basictests.cpp
Comment (pauldreik) [reply tests/ondemand/ondemand_basictests.cpp]: I undid the lambda extraction.

**Comment by pauldreik** [comment]
I undid the lambda extraction in all places, so what remains is the loop fix and the cleanup of cstyle cast, unsigned/signed comparisons etc.

It would be great if we could use clang format to avoid messing up the formatting.

**Review by jkeiser — APPROVED** [review]
This looks good; thank you!!

Per 0x800, @lemire might know more--it's not a magic constant I spent time with. The reason the ondemand number tests look like they were written by me is I copy/pasted them from basictests :) (Not that I haven't refactored them over time.)

I'm all for being able to clang format; we have an issue to figure out precisely what coding style we want to enforce, I believe, unless we resolved that and went with the Google one. I can't recall right now :)

File: tests/ondemand/ondemand_basictests.cpp
Comment (pauldreik) [reply tests/ondemand/ondemand_basictests.cpp:213]: Fixing compiler warnings that turned into a refactor - wanted to keep what is essentially a "double2string()" function outside of the test loop, so it was easier to see what the test loop did.

**Comment by pauldreik** [comment]
The CI fuzzer error is real and unrelated to this PR. Trying to understand it.

**Comment by pauldreik** [comment]
> The CI fuzzer error is real and unrelated to this PR. Trying to understand it.

The issue is reported in #1273

File: tests/number_comparison.h
Comment (lemire) [inline tests/number_comparison.h:13]: Do we ever use f32_ulp_dist?

**Comment by lemire** [comment]
@pauldreik I suggest we retire entirely the ULP functions.

The path you refer to, with a constant, was never taken and if it was ever taken, all that we needed is that it generated a large value. The logic of this function was that it would give a small value corresponding to the ULP, otherwise it might as well give any large value, we did not care. I also wanted it to be fast.

To put it in different words, this function is incorrect when the ULP is large... it generates a large value which may not be the actual ULP, except when the ULP is very small.  The path never taken correspond to a sign difference. Well, if you have a sign difference, then I don't care about how big the difference is, we just want failure.



But we don't need ULP at all now. Except for systems that does not have exact floats, we always require exact results. So we can just do "x == y" instead of "ulp(x,y)==0". If there is an error, we don't want to print out an ULP, we just want to bail out... because we want the results to be exact.

This is legacy code.

File: tests/number_comparison.h
Comment (pauldreik) [reply tests/number_comparison.h:13]: appearantly not!

**Comment by pauldreik** [comment]
> @pauldreik I suggest we retire entirely the ULP functions.

I like it! Got rid of all of it. Tests go through locally, waiting for green CI and then merge.

**Pull request merged**
